frontend/js/src/artist-similarity/src/Data.tsx

```diff
@@ -0,0 +1,168 @@
+import React, {useState, useEffect} from "react";
+import Graph from "./Graph";
+import Input from "./Input";
+
+type ColorType = {
+    red: number;
+    green: number;
+    blue: number;
+    opacity: number;
+    mixed: string;
+}
+
+type ArtistType = {
+    artist_mbid: string;
+    name: string;
+    comment: string;
+    type: string;
+    gender: string;
+    score?: number;
+    reference_mbid?: string;
+}
+    
+type MarkupResponseType = {
+    data: string;
+    type: "markup";
+}
+
+type DatasetResponseType = {
+    columns: Array<string>;
+    data: Array<ArtistType>;
+    type: "dataset";
+}
+
+type ApiResponseType = MarkupResponseType | DatasetResponseType;
+
+const Color = (red: number, green: number, blue: number): ColorType => {
+    return {
+        red: red,
+        green: green,
+        blue: blue,
+        opacity: 1.0,
+        mixed: "rgba(" + red + "," + green + "," + blue + "," + 1.0 + ")",
+    }
+}
+
+const changeOpacity = (color: ColorType, opacity: number): ColorType => {
+    return {
+        red: color.red,
+        green: color.green,
+        blue: color.blue,
+        opacity: opacity,
+        mixed: "rgba(" + color.red + "," + color.green + "," + color.blue + "," + opacity + ")",
+    }
+}
+
+const colorGenerator = (): ColorType => {
+    var red = Math.floor(Math.random() * 200);
+    var green = Math.floor(Math.random() * 200);
+    var blue  = Math.floor(Math.random() * 200);
+
+    return Color(red, green, blue);
+}
+
+const mixColor = (color1: ColorType, color2: ColorType, weight: number): ColorType => {
+    var red = Math.floor((color2.red - color1.red) * weight + color1.red);
+    var green = Math.floor((color2.green - color1.green) * weight + color1.green);
+    var blue = Math.floor((color2.blue - color1.blue) * weight + color1.blue);          
+    var opacity = Math.floor((color2.opacity - color1.opacity) * weight + color1.opacity);
+
+    return Color(red, green, blue);
+}
+
+const Data = () => {
+    const LIMIT_VALUE = 18;
+    const BASE_URL = "https://labs.api.listenbrainz.org/similar-artists/json?algorithm=session_based_days_7500_session_300_contribution_5_threshold_10_limit_100_filter_True_skip_30&artist_mbid=";
+    
+    var artist_mbid = "8f6bd1e4-fbe1-4f50-aa9b-94c450ec0f11";
```

Comment:
I think it would make sense to store the artist MBID in this component's state, and pass the `setArtistMBID` function down to the children components instead of passing fetchData. Then you can react to changes to `artistMBID` in this parent component.

In short:
```
const [artistMBID, setArtistMBID] = useState("8f6bd1e4-fbe1-4f50-aa9b-94c450ec0f11");
[…]
// On line 107
useEffect(() => {
        fetchData(artistMBID);
}, [artistMBID]);
[…]
// In the return condition
<Input onArtistChange={setArtistMBID} …
<Graph onArtistChange={setArtistMBID} …
```

Reply:
Implemented your suggestion